plookup/src/table/hash_tables/tables.rs

```diff
@@ -0,0 +1,207 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+///! In circuit hashing can be conducted using table lookups from the
+///! tables defined in this file.
+/// This currently assumes that T_S = 4. It is possible that in future T_S > 4,
+/// in which case this file will have to be adjusted. T_S is the arity of the
+/// vectors.
+use crate::table::hash_tables::constants::{N, S, T_S, V};
+use dusk_plonk::fft::Polynomial;
+use dusk_plonk::prelude::BlsScalar;
+use std::convert::TryInto;
+
+/// A HashTable consists of three different segments, each of arity
+/// vectors containing Bls Scalars. These values are precomputed and
+/// stored. These are the retrieved in the lookup functions and placed
+/// into seperate structs for the proof.
+/// A vector x in (F_p)^t goes through r rounds of a round function R.
+/// The result is another vector y in (F_p)^t.
+#[derive(Debug)]
+pub struct HashTable {
+    pub first_rows: Vec<[BlsScalar; 4]>,
+    pub middle_rows: Vec<[BlsScalar; 4]>,
+    pub end_rows: Vec<[BlsScalar; 4]>,
+}
+
+impl HashTable {
+    /// Create a new hash table and makes
+    /// empty vectors for each field.
+    pub fn new() -> Self {
+        Self {
+            first_rows: vec![],
+            middle_rows: vec![],
+            end_rows: vec![],
+        }
+    }
+
+    // The whole lookup table will be constructed in 3 parts: the first rows where the third
+    // entry is derived from the function F, i.e. the rows are of the form (_, _, F(i), ...).
+    // The middle rows are where the first entries are between V+1 and s_i for some i.
+    // The binary rows are at the bottom of the table, and they enumerate all binary possibilities
+    // on T_S bits.
+    fn f_rows(&mut self, f: Polynomial) {
+        for i in 0..(V + 1) {
+            let eval = f.evaluate(&BlsScalar::from(i as u64));
+            let row = [
+                BlsScalar::from(i as u64),
+                BlsScalar::zero(),
+                eval,
+                -BlsScalar::one(),
+            ];
+
+            self.first_rows.push(row);
```

Comment:
If you'll always push `V + 1` elements. This does not need to be a `Vec` and can be an array instead.